### PR TITLE
GH-413 - Describe what's in the block and why

### PR DESCRIPTION
**Pull request by chb0github**
Documenting blockchain structs and functions

### CODE REVIEW

File: src/blockchain/src/block.rs [hunk 1]
```diff
@@ -75,7 +79,16 @@ impl Block {
     }
 }
 
-// struct Transaction define the details of a transaction in a block
+/// ## Transactions ##
+/// A Transaction consists of
+/// - `trans_type` Pretty obvious. Used to help decode the rest of the transaction
```
Comment (prince-chrismc) [inline src/blockchain/src/block.rs]: These should go over the fields

File: src/blockchain/src/block.rs [hunk 2]
```diff
@@ -75,7 +79,16 @@ impl Block {
     }
 }
 
-// struct Transaction define the details of a transaction in a block
+/// ## Transactions ##
```
Comment (prince-chrismc) [inline src/blockchain/src/block.rs]: Did you check the rendered output... see's odd to repeat the name

bad markdown

```suggestion
/// ## Transaction
```

File: src/blockchain/src/blockchain.rs [hunk 3]
```diff
@@ -60,12 +61,19 @@ impl Default for Config {
 }
 
 #[derive(Serialize, Deserialize)]
+/// This is our blockchain. It consists of the blocks and observers of transactions and blocks
 pub struct Blockchain {
     #[serde(skip)]
-    // this should actually be a Map<Transaction,Vec<OnTransactionSettled>> but that's later
+    /// This is where we keep our transaction observers.
+    /// Currently there can only be one, but this is incorrect. This should actually be a
+    /// Map<Transaction,Vec<OnTransactionSettled>> but that's later
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs]: should not be in the docs... just a regular comment

File: src/blockchain/src/blockchain.rs [hunk 4]
```diff
@@ -111,6 +119,15 @@ impl Blockchain {
...
     }
+    /// Submit a transaction here. When the transaction is settled, you will be notified
+    /// # Arguments
+    /// * `trans` the transaction to be submitted
+    /// * `on_done` the callback that will be invoked when the transaction settles
+    ///
+    /// # Example
+    ///         chain.submit_transaction(transaction.clone(), move |t: Transaction| {
+    ///             println!("transaction {} settled",t);
+    ///         })
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs]: should be an actual code snippet so it gets ran with the doc  tests

File: src/blockchain/src/blockchain.rs [hunk 5]
```diff
@@ -119,14 +136,30 @@ impl Blockchain {
...
         self
     }
-
+    /// This will let all listeners know that the transaction has settled. This shouldn't be accessible
+    /// outside of the crate or probably by any other receive than the blockchain
+    /// # Arguments
+    /// * `trans` the transaction that has settled
+    ///
+    /// # Example
+    ///         chain.notify_transaction_settled(transaction);
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs]: actually snippet please

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
rustdoc needs work 🙏

File: src/blockchain/src/blockchain.rs (hunk 5 above)
Comment (chb0github) [reply src/blockchain/src/blockchain.rs]: Oh, I think I get what you mean

File: src/blockchain/src/block.rs (hunk 1 above)
Comment (chb0github) [reply src/blockchain/src/block.rs]: remember I was trying to discuss standardization on this? And you were like: "It should all just self-doc and we don't need a standard"

I'll do it, as I have no opinion on the matter

Comment (chb0github) [reply src/blockchain/src/block.rs]: I am going to take this a step further. Ironically, we've outsourced our code formatting to a group of people we've never met and had no say in the process, but the documentation is totally freestyle. 

I am literally going to rip off the format in the rust core libs

Comment (prince-chrismc) [reply src/blockchain/src/block.rs]: your comments made no sense... it's already specd on how it should be written

https://doc.rust-lang.org/rust-by-example/meta/doc.html

https://doc.rust-lang.org/rustdoc/index.html

File: src/blockchain/src/block.rs [hunk 6]
```diff
@@ -48,11 +48,12 @@ pub fn get_publickey_from_keypair(
...
 
-// struct Block define a block strcuture
 #[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 pub struct Block {
     pub header: Header,
+    /// `transactions` The transactions that make up this block
     pub transactions: Vec<Transaction>,
+    /// `signature` a signature of the whole block - it's header and transactions
     pub signature: BlockSignature,
 }
```
Comment (prince-chrismc) [inline src/blockchain/src/block.rs]: You dont need to duplicate the field name

```suggestion
    pub header: Header,
    /// The list of transactions that make up a block
    pub transactions: Vec<Transaction>,
    /// A signature of the entire block - it's header and transactions
    pub signature: BlockSignature,
}
```

**Comment by prince-chrismc** [comment]
You should try `cargo doc` to see the output

File: src/blockchain/src/block.rs (hunk 1 above)
Comment (chb0github) [reply src/blockchain/src/block.rs]: I don't see the part about how structs should be documented. But, it's fine. I have implemented as you requested.

File: src/blockchain/src/block.rs [hunk 7]
```diff
@@ -78,19 +78,19 @@ impl Block {
 
 #[derive(Serialize, Deserialize, Debug, Clone, Hash, PartialEq, Eq)]
 pub struct Transaction {
-    /// - `trans_type` Pretty obvious. Used to help decode the rest of the transaction
+    /// Pretty obvious. Used to help decode the rest of the transaction
```
Comment (prince-chrismc) [inline src/blockchain/src/block.rs]:
```suggestion
    /// Used to help decode the rest of the transaction
```

File: src/blockchain/src/block.rs [hunk 8]
```diff
@@ -78,19 +78,19 @@ impl Block {
...
+    /// User payload no decoding will be supplied
     pub payload: Vec<u8>,
-    /// - `nonce` A random value added to the transaction payload to harden it
+    /// A random value added to the transaction payload to harden it
     nonce: u128, // Adds a salt to harden
-    /// - `transaction_hash` A hash of all the fields in the transaction
+    /// A hash of all the fields in the transaction
     pub transaction_hash: HashDigest,
-    /// - `signature` The signature of the transaction_hash data in this transaction
+    /// The signature of the transaction_hash data in this transaction
```
Comment (prince-chrismc) [inline src/blockchain/src/block.rs]: I was reading online there a way to link to other fields, could you mind playing with that?

Comment (chb0github) [reply src/blockchain/src/block.rs]: Let me see

File: docs/blockchain/whats_in_the_blockchain.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+# Unresolved questions
+
+None.
+
+# Future possibilities
+
+This structure is most definitely going to change as we learn and grow. There are already several outstanding tickets to do so
+
+## A structural breakdown of our blockchain ##
```
Comment (prince-chrismc) [inline docs/blockchain/whats_in_the_blockchain.md:45]:
```suggestion
## A structural breakdown of our blockchain
```

File: docs/blockchain/whats_in_the_blockchain.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+# Prior art
+
+When constructing this blockchain we consulted the structure of several famous blockchain implementations. Namely
+`BTC` and `ETH`. Bitcoin was the simplest, but it has the worst growing pains and doesn't handle chain merging well.
+Etherium manages chain merging by treating it not as a blockchain, but as a block tree - `Merkle tree`. In addition, all of the
+new implementations of blockchain allow for a list of transactions in a single block. This keep the block count lower and thus
+increases overall efficiency since it's the blocks that require consensus and take time.
+
+TO that end, our blockchain contains:
```
Comment (prince-chrismc) [inline docs/blockchain/whats_in_the_blockchain.md:32]:
```suggestion
To that end, our blockchain contains:
```

File: docs/blockchain/whats_in_the_blockchain.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+## A structural breakdown of our blockchain ##
+
+Our blockchain consists
+
+* series of blocks where
+    * each block has
+        * a Header
+            * which contains it's parent hash. This is the part of the block chain that makes it a `chain`. Each block has a reference to it's parent going all the way back to the genesis. Basically an inductive proof.
+            * The block committers address. On any block chain you need to identify the participant. The address isn't really any kind of routing destination or origin. It's the unique id of a participant and nothing more.
+            * ~the transactions root - This is an optimization to verify all transactions in a block. We can simply add the transaction hash (256bit Keccak Hash of the root node of Transactions Merkle tree). Certain block chain implementations aren't blocks, they are actually merkle trees (this is done to handle chain merging conflicts)~
```
Comment (prince-chrismc) [inline docs/blockchain/whats_in_the_blockchain.md:54]: this was revived

**Pull request merged**